File: PreternaturalProject/src/main/java/com/preternatural/PreternaturalProject/services/CharacterService.java
package com.preternatural.PreternaturalProject.services;

import java.util.List;
import java.util.Optional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.preternatural.PreternaturalProject.models.Character;
import com.preternatural.PreternaturalProject.models.CharacterRole;
import com.preternatural.PreternaturalProject.models.CharacterRole2;
import com.preternatural.PreternaturalProject.models.Species;
import com.preternatural.PreternaturalProject.models.Stage;
import com.preternatural.PreternaturalProject.repositories.CharacterRepository;

@Service("characterService")
public class CharacterService {

	private CharacterRepository characterRepository;
	
	@Autowired
	public void setCharacterRepository(CharacterRepository characterRepository) {
		this.characterRepository = characterRepository;
	}
	
	public List<Character> getAllCharacters() {
		return characterRepository.findAll();
	}
	
	public Character getCharacterById(int id) {
		Optional<Character> p = characterRepository.findById(id);
		if (p.isPresent()) {
			return p.get();
		} else {
			return null;
		}
	}
	
	public void deleteCharacterById(int id) {
		characterRepository.deleteById(id);
	}
	
	public List<Character> getCharactersByFirstname(String firstname) {
		return characterRepository.getCharactersByFirstname(firstname);
	}
	
	public List<Character> getCharactersByLastname(String lastname) {
		return characterRepository.getCharactersByLastname(lastname);
	}
	
	public List<Character> getCharactersByRole(CharacterRole role) {
		return characterRepository.getCharactersByRole(role);
	}
	
	public List<Character> getCharactersByRole2(CharacterRole2 role2) {
		return characterRepository.getCharactersByRole2(role2);
	}
	
	public List<Character> getCharactersByAge(int age) {
		return characterRepository.getCharactersByAge(age);
	}
	
	
	public List<Character> getCharactersBySpecies(Species species) {
		return characterRepository.getCharactersBySpecies(species);
	}
	public List<Character> getCharactersByStage(Stage stage) {
		return characterRepository.getCharactersByStage(stage);
	}
	
	public void createCharacter(Character character) {
		characterRepository.save(character);
	}
	
	public int updateCharacter(Character character) {
		Character p = characterRepository.save(character);
		if (p != null) {
			return p.getId();
		} else {
			return 0;
		}
	}
	
}

File: Preternatural-Angular/src/app/services/character.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Character } from '../models/character.model';

@Injectable({
  providedIn: 'root'
})



export class CharacterService {

  apiUrl: string = "http://localhost:8085/characters";
  
  constructor(private httpClient: HttpClient) { 
    
    
  }


  public getAllCharacters(){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/all`);
  }

  public getCharacterByFirstName(firstname: string){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getByFirstname/`+ firstname);
  }

  public getCharactersByLastName(lastname: string){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getByLastname/`+ lastname);
  }

  public getCharactersByRole(role: string){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getByRole/`+ role);
  }

  public getCharactersByRole2(role2: string){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getByRole2/`+ role2);
  }

  public getCharactersByAge(age: number){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getByAge/`+ age);
  }

  public getCharactersBySpecies(species: string){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getBySpecies/`+ species);
  }

  public getCharactersByStage(stage: string){
    return this.httpClient.get<Character[]>(`${this.apiUrl}/getByStage/`+ stage);
  }

  public createCharacter(character: Character){
    return this.httpClient.post(`${this.apiUrl}/create`, character);
  }
  
  public updateCharacter(character: Character){
    return this.httpClient.put(`${this.apiUrl}/update`, character);
  }
}
